File: 백준/Gold/14466. 소가 길을 건너간 이유 6/소가 길을 건너간 이유 6.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Main {
    static int[] xd = {-1, 0, 1, 0};
    static int[] yd = {0, 1, 0, -1};
    static int N, K, R;
    static boolean[][][] linked;
    static List<int[]> cows = new ArrayList<>();
    static boolean[][] visit;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        K = Integer.parseInt(st.nextToken());
        R = Integer.parseInt(st.nextToken());
        linked = new boolean[N + 1][N + 1][4];
        //길 입력
        for (int i = 0; i < R; i++) {
            st = new StringTokenizer(br.readLine());
            int x1 = Integer.parseInt(st.nextToken());
            int y1 = Integer.parseInt(st.nextToken());
            int x2 = Integer.parseInt(st.nextToken());
            int y2 = Integer.parseInt(st.nextToken());
            for (int dir = 0; dir < 4; dir++) {
                int nx = x1 + xd[dir];
                int ny = y1 + yd[dir];
                if (nx == x2 && ny == y2) {
                    linked[x1][y1][dir] = true;
                    linked[x2][y2][(dir + 2) % 4] = true;
                }
            }
        }
        //소 위치 입력
        for (int i = 0; i < K; i++) {
            st = new StringTokenizer(br.readLine());
            int x = Integer.parseInt(st.nextToken());
            int y = Integer.parseInt(st.nextToken());
            cows.add(new int[]{x, y});
        }

        int answer = 0;
        for (int i = 0; i < K; i++) {
            visit = new boolean[N + 1][N + 1];
            bfs(cows.get(i)[0], cows.get(i)[1]);
            for (int j = i + 1; j < K; j++) {
                if (!visit[cows.get(j)[0]][cows.get(j)[1]])
                    answer++;
            }
        }
        System.out.println(answer);
    }

    static void bfs(int a, int b) {
        Queue<int[]> q = new LinkedList<int[]>();
        q.add(new int[]{a, b});
        visit[a][b] = true;
        //길을 건너지 않고 만날 수 있는 소 확인
        while (!q.isEmpty()) {
            int[] cur = q.poll();
            int x = cur[0];
            int y = cur[1];
            for (int i = 0; i < 4; i++) {
                //길이 있는 경우 패스
                if (linked[x][y][i])
                    continue;

                int nx = x + xd[i];
                int ny = y + yd[i];
                if (isInRange(nx, ny) && !visit[nx][ny]) {
                    visit[nx][ny] = true;
                    q.add(new int[]{nx, ny});
                }
            }
        }
    }

    static boolean isInRange(int x, int y) {
        return x >= 1 && x <= N && y >= 1 && y <= N;
    }
}

File: 백준/Gold/14466. 소가 길을 건너간 이유 6/소가 길을 건너간 이유 6.py
def bfs(cow):
    q = []
    q.append(cow)
    visit[cow[0]][cow[1]] = True
    while q!=[]:
        cur = q.pop(0)
        for i in range(4):
            if linked[cur[0]][cur[1]][i]:
                continue
                
            nx = cur[0]+xd[i]
            ny = cur[1]+yd[i]
            if 1<=nx<=n and 1<=ny<=n and not visit[nx][ny]:
                visit[nx][ny] = True
                q.append([nx,ny])

xd = [1,0,-1,0]
yd = [0,1,0,-1]

n,k,r = map(int, input().split())
linked = [[[False]*4 for _ in range(n+1)] for _ in range(n+1)]
cows = []
for _ in range(r):
    x1,y1,x2,y2 = map(int, input().split())
    for i in range(4):
        nx = x1+xd[i]
        ny = y1+yd[i]
        if [nx,ny]==[x2,y2]:
            linked[x1][y1][i] = True
            linked[x2][y2][(i+2)%4] = True
for _ in range(k):
    x,y = map(int, input().split())
    cows.append([x,y])
# print(linked)
# print(cows)

answer = 0
visit = [[False]*(n+1) for _ in range(n+1)]
for i in range(k):
    visit = [[False]*(n+1) for _ in range(n+1)]
    bfs(cows[i])
    for j in range(i+1,k):
        if not visit[cows[j][0]][cows[j][1]]:
            answer+=1
print(answer)